### PR TITLE
Fix `--port=0` option to correctly open on random port

### PR DESCRIPTION
**Pull request by GregBrimble**
Dependent on https://github.com/cloudflare/miniflare/pull/382

### CODE REVIEW

**Comment by rozenmd** [comment]
Hey so with https://github.com/cloudflare/miniflare/pull/382 merged, is this good to go?

**Comment by GregBrimble** [comment]
Just waiting on a Miniflare release 😊

**Comment by rozenmd** [comment]
Also may want to add this to `--experimental-local`

**Comment by rozenmd** [comment]
@mrbbot pls

**Comment by GregBrimble** [comment]
https://github.com/cloudflare/wrangler2/pull/2003

**Review by rozenmd — APPROVED** [review]

**Comment by GregBrimble** [comment]
hmm, this actually isn't quite ready. It improves things, but it doesn't update the hotkey things, so pressing `b` will try to open `localhost:0` :(

File: packages/wrangler/src/dev/dev.tsx [hunk 1]
```diff
@@ -171,25 +171,39 @@ export function DevImplementation(props: DevProps): JSX.Element {
 	);
 }
 
+let ip: string;
```
Comment (GregBrimble) [inline packages/wrangler/src/dev/dev.tsx:174]: Sheer madness, I know.

Comment (GregBrimble) [reply packages/wrangler/src/dev/dev.tsx:174]: The problem is that `useHotkeys()` relies on these values for it's `b` "open a browser" feature. If we store this in proper React state, it gets the value fine, but we then need to call `setIp` or whatever in the `onReady` callback. This causes the entire component to re-render, which re-initializes Miniflare, which causes the `onReady` callback to be called again, etc. etc.

I tried an `if (newValue !== newValue) { setValue() }`, but still couldn't prevent a render loop.

File: packages/wrangler/src/dev.tsx [hunk 2]
```diff
@@ -467,11 +467,13 @@ export async function startDev(args: StartDevOptions) {
 					accountId={configParam.account_id || getAccountFromCache()?.id}
 					assetPaths={assetPaths}
 					assetsConfig={configParam.assets}
-					port={args.port || configParam.dev.port || (await getLocalPort())}
-					ip={args.ip || configParam.dev.ip}
+					initialPort={
+						args.port ?? configParam.dev.port ?? (await getLocalPort())
```
Comment (GregBrimble) [inline packages/wrangler/src/dev.tsx:471]: I'm going to come back and refactor `getLocalPort` and `waitForPortToBeAvailable` so they're no longer race-y in a subsequent PR.

File: packages/wrangler/src/dev/start-server.ts [hunk 3]
```diff
@@ -325,14 +325,11 @@ export async function startLocalServer({
...
 		// port for the worker
-
-		if (port !== 0) {
-			await waitForPortToBeAvailable(port, {
-				retryPeriod: 200,
-				timeout: 2000,
-				abortSignal: abortController.signal,
-			});
-		}
+		await waitForPortToBeAvailable(initialPort, {
```
Comment (rozenmd) [inline packages/wrangler/src/dev/start-server.ts:328]: if initialPort is 0, we don't have to wait for the port available though - node should just give us an available one?

File: packages/wrangler/src/proxy.ts [hunk 4]
```diff
@@ -651,6 +651,11 @@ export async function waitForPortToBeAvailable(
 		}
 
 		function checkPort() {
+			if (port === 0) {
```
Comment (rozenmd) [inline packages/wrangler/src/proxy.ts:654]: ...oh okay

**Review by rozenmd — APPROVED** [review]
I had a question, it got answered later in the PR

LGTM 👍

File: packages/wrangler/src/proxy.ts (hunk 4 above)
Comment (GregBrimble) [reply packages/wrangler/src/proxy.ts:654]: Yup. I'll tidy this up in a subsequent PR removing these race-y functions.

**Pull request merged**